compiler/src/dotty/tools/dotc/ast/Trees.scala

```diff
@@ -1595,7 +1595,7 @@ object Trees {
               cpy.DefDef(tree)(name, transformParamss(paramss), transform(tpt), transform(tree.rhs))
             case tree @ TypeDef(name, rhs) =>
               cpy.TypeDef(tree)(name, transform(rhs))
-            case tree @ Template(constr, parents, self, _) if tree.derived.isEmpty =>
```

Comment:
The guard made sure nothing was forgotten in the transform. Instead of a guard it could be an assert. But then we could not fill in the missing case (where a `derived` is not empty) in `transformMoreCases`. 

So I think overall it's best to leave it in. Maybe add a comment.

Reply:
I did test it locally with an assert and then removed it. I could add back the assert with a meaningful message. It might be clearer this way that the `Template` was not supposed to have a `derived` set.

Are we only supposed to have non-empty `derived` for untyped templates? Or are there cases where the typed template is expected to have something in `derived`?

Reply:
We don't know. As you tested, there are no occurrences where derived is empty when we transform. But that is not necessarily true for the future. With the original scheme we fall through into `transformMoreCases` when we see such a tree where we would fail unless `transformMoreCases` explicitly provides for it. I think that's what we want. An `assert` is worse since it prevents the handling of the case. The chance is high that if such a case arises in the future, and we hit the `assert`, we just remove it, since we have now found a legitimate case where it mis-fires. But that would be bad since then other maps would have a hole where they silently drop the derived field. So I really think the original code is the best possible design.